src/recurrent/rnn-time-step.js

```diff
@@ -0,0 +1,169 @@
+   *
+   * @param {Number[]} input
+   * @returns {number}
+   */
+  runInput(input) {
+    this.runs++;
+    let model = this.model;
+    let errorSum = 0;
+    let equation;
+    while (model.equations.length < input.length - 1) {
+      this.bindEquation();
+    }
+    const outputs = [];
+
+    if (this.inputSize === 1) {
+      for (let inputIndex = 0, max = input.length - 1; inputIndex < max; inputIndex++) {
+        // start and end tokens are zeros
+        equation = model.equations[inputIndex];
+
+        const current = input[inputIndex];
+        const next = input[inputIndex + 1];
+        const output = equation.runInput([current]);
+        for (let i = 0; i < output.weights.length; i++) {
+          const error = output.weights[i] - next;
+          // set gradients into log probabilities
+          errorSum += Math.abs(error);
+
+          // write gradients into log probabilities
+          output.deltas[i] = error;
+          outputs.push(output.weights);
+        }
+      }
+    } else {
+      for (let inputIndex = 0, max = input.length - 1; inputIndex < max; inputIndex++) {
+        // start and end tokens are zeros
+        equation = model.equations[inputIndex];
+
+        const current = input[inputIndex];
+        const next = input[inputIndex + 1];
+        const output = equation.runInput(current);
+        for (let i = 0; i < output.weights.length; i++) {
+          const error = output.weights[i] - next[i];
+          // set gradients into log probabilities
+          errorSum += Math.abs(error);
+
+          // write gradients into log probabilities
+          output.deltas[i] = error;
+          outputs.push(output.weights);
+        }
+      }
+    }
+    //this.model.equations.length - 1;
+    this.totalCost = errorSum;
+    return errorSum;
+  }
+
+  runBackpropagate() {
+    for (let i = this.model.equations.length - 1; i > -1; i--) {
+      this.model.equations[i].runBackpropagate();
+    }
+  }
+
+
+  /**
+   *
+   * @param {Number[]|Number} [input]
+   * @param {Number} [maxPredictionLength]
+   * @param {Boolean} [isSampleI]
+   * @param {Number} temperature
+   * @returns {Number[]|Number}
+   */
+  run(input = [], maxPredictionLength = 1, isSampleI = false, temperature = 1) {
+    if (!this.isRunnable) return null;
+    const model = this.model;
+    while (model.equations.length < maxPredictionLength) {
+      this.bindEquation();
+    }
+    let lastOutput;
+    if (this.inputSize === 1) {
+      for (let i = 0; i < input.length; i++) {
+        let outputMatrix = model.equations[i].runInput([input[i]]);
+        lastOutput = outputMatrix.weights;
+      }
+    } else {
+      for (let i = 0; i < input.length; i++) {
+        let outputMatrix = model.equations[i].runInput(input[i]);
+        lastOutput = outputMatrix.weights;
+      }
+    }
+    if (this.outputSize === 1) {
+      return lastOutput[0]
+    }
+    return lastOutput;
+  }
+
+  /**
+   *
+   * @returns {Function}
+   */
+  toFunction() {
```

Comment:
Thank you for including this!! I love that it's stubbed out and throws, so it is obviously clear if ever used what went wrong, but when we are ready to implement it we already have it stubbed out and just need to finish it.

Reply:
Nice.